it/src/main/java/com/google/cloud/teleport/it/spanner/DefaultSpannerResourceManager.java

```diff
@@ -0,0 +1,244 @@
+  }
+
+  private synchronized void createInstance() {
+    LOG.info("Creating instance {} in project {}.", instanceId, projectId);
+    InstanceInfo instanceInfo =
+        InstanceInfo.newBuilder(InstanceId.of(projectId, instanceId))
+            .setInstanceConfigId(InstanceConfigId.of(projectId, region))
+            .setDisplayName(instanceId)
+            .setNodeCount(1)
+            .build();
+    try {
+      instanceAdminClient.createInstance(instanceInfo).get();
+      hasInstance = true;
+      LOG.info("Successfully created instance {}.", instanceId);
+    } catch (ExecutionException | InterruptedException e) {
+      throw new SpannerResourceManagerException("Failed to create instance.", e);
+    }
+  }
+
+  private synchronized void createDatabase() {
+    LOG.info("Creating database {} in instance {}.", databaseId, instanceId);
+    try {
+      databaseAdminClient.createDatabase(instanceId, databaseId, Collections.emptyList()).get();
+      databaseClient = spanner.getDatabaseClient(DatabaseId.of(projectId, instanceId, databaseId));
+      LOG.info("Successfully created database {}.", databaseId);
+    } catch (ExecutionException | InterruptedException e) {
+      throw new SpannerResourceManagerException("Failed to create database.", e);
+    }
+  }
+
+  private void checkIsUsable() throws IllegalStateException {
+    if (spanner.isClosed()) {
+      throw new IllegalStateException("Manager has cleaned up all resources and is unusable.");
+    }
+  }
+
+  private void checkHasDatabase() throws IllegalStateException {
+    if (databaseClient == null) {
+      throw new SpannerResourceManagerException("Database does not exist");
+    }
+  }
+
+  @Override
+  public synchronized void createTable(String statement) throws IllegalStateException {
+    checkIsUsable();
+    if (!hasInstance) {
+      createInstance();
+    }
+    if (databaseClient == null) {
+      createDatabase();
+    }
+    LOG.info("Creating table in database {} using statement '{}'", databaseId, statement);
+    try {
+      databaseAdminClient
+          .updateDatabaseDdl(
+              instanceId, databaseId, ImmutableList.of(statement), /* operationId= */ null)
+          .get();
+      LOG.info("Successfully created table in database {}.", databaseId);
+    } catch (ExecutionException | InterruptedException e) {
+      throw new SpannerResourceManagerException("Failed to create table.", e);
+    }
+  }
+
+  @Override
+  public synchronized void insertTableRecord(Mutation tableRecord) throws IllegalStateException {
+    insertTableRecords(ImmutableList.of(tableRecord));
+  }
+
+  @Override
+  public synchronized void insertTableRecords(Iterable<Mutation> tableRecords)
+      throws IllegalStateException {
+    String tableId = Iterables.get(tableRecords, 0).getTable();
+    LOG.info(
+        "Inserting {} records into table {}.{}.{}",
+        Iterables.size(tableRecords),
+        instanceId,
+        databaseId,
+        tableId);
+
+    checkIsUsable();
```

Comment:
Maybe move these checks before the log? It's less likely to give off the impression the write was attempted.